subprojects/gradle/build-trace/src/main/kotlin/com/avito/android/build_trace/internal/critical_path/CriticalPathListener.kt

```diff
@@ -0,0 +1,47 @@
+package com.avito.android.build_trace.internal.critical_path
+
+import com.avito.android.gradle.metric.AbstractBuildEventsListener
+import com.avito.android.gradle.profile.BuildProfile
+import com.avito.android.gradle.profile.TaskExecution
+import com.avito.graph.ShortestPath
+import org.gradle.BuildResult
+import org.gradle.api.Task
+
+internal class CriticalPathListener(
+    private val serialization: CriticalPathSerialization
+) : AbstractBuildEventsListener(), CriticalPathProvider {
+
+    private val operations = mutableSetOf<TaskOperation>()
+
+    private val criticalPath: List<TaskOperation> by lazy {
+        ShortestPath(operations)
+            .find()
+            .map { it.invertTime() }
+    }
+
+    override fun path(): List<TaskOperation> {
+        return criticalPath
+    }
+
+    // Using a TaskExecutionListener instead of OperationCompletionListener
+    //   due to https://github.com/gradle/gradle/issues/15824
+    override fun afterExecute(task: Task, state: TaskExecution) {
```

Comment:
This implementation will be a problem in two cases:

- To share critical path between multiple plugins
- For configuration cache

I'll find a solution in another issue\PR.